### PR TITLE
Single binary event controller

### PR DESCRIPTION
**Pull request by themue**
**Description**

Changes proposed in this pull request:

- Creation of only one eventing-controller binary
- Switch between used backends by usage of environment variable `BACKEND`
- Environment variable is part of the deployments in the helm charts and set in the global values

**Related issue(s)**

See also #10801

### CODE REVIEW

File: components/eventing-controller/cmd/eventing-controller/main.go [hunk 1]
```diff
@@ -1,89 +1,77 @@
...
-var scheme = runtime.NewScheme()
+const (
+	ENV_BACKEND = "BACKEND"
 
-func init() {
-	_ = clientgoscheme.AddToScheme(scheme)
-	_ = eventingv1alpha1.AddToScheme(scheme)
-	_ = apigatewayv1alpha1.AddToScheme(scheme)
+	ENV_BACKEND_VALUE_BEB  = "BEB"
+	ENV_BACKEND_VALUE_NATS = "NATS"
```
Comment (marcobebway) [inline components/eventing-controller/cmd/eventing-controller/main.go:21]: Does it make sense to not export those constants outside of the main package?

File: components/eventing-controller/cmd/eventing-controller/nats/nats.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	eventingv1alpha1 "github.com/kyma-project/kyma/components/eventing-controller/api/v1alpha1"
+	"github.com/kyma-project/kyma/components/eventing-controller/pkg/application"
+	"github.com/kyma-project/kyma/components/eventing-controller/pkg/env"
+	subscription "github.com/kyma-project/kyma/components/eventing-controller/reconciler/subscription-nats"
+)
+
+// Commander implementes the Commander interface.
```
Comment (marcobebway) [inline components/eventing-controller/cmd/eventing-controller/nats/nats.go:22]: Typo `implementes`

File: components/eventing-controller/cmd/eventing-controller/beb/beb.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	apigatewayv1alpha1 "github.com/kyma-incubator/api-gateway/api/v1alpha1"
+	eventingv1alpha1 "github.com/kyma-project/kyma/components/eventing-controller/api/v1alpha1"
+	"github.com/kyma-project/kyma/components/eventing-controller/pkg/application"
+	"github.com/kyma-project/kyma/components/eventing-controller/pkg/env"
+	"github.com/kyma-project/kyma/components/eventing-controller/reconciler/subscription"
+)
+
+// Commander implementes the Commander interface.
```
Comment (marcobebway) [inline components/eventing-controller/cmd/eventing-controller/beb/beb.go:23]: Typo `implementes`

File: components/eventing-controller/cmd/eventing-controller/nats/nats.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+// Start implements the Commander interface and starts the manager.
+func (c *Commander) Start() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	dynamicClient := dynamic.NewForConfigOrDie(c.restCfg)
+	applicationLister := application.NewLister(ctx, dynamicClient)
+
+	// TODO Harmonization. Not subscription and subscription-nats.
```
Comment (marcobebway) [inline components/eventing-controller/cmd/eventing-controller/nats/nats.go:79]: Is there a github issue to take care of this? If yes, can you please link it here.

File: components/eventing-controller/cmd/eventing-controller/main.go [hunk 5]
```diff
@@ -1,89 +1,77 @@
...
+	if err := commander.Init(); err != nil {
+		logger.Error(err, "unable to start manager")
 		os.Exit(1)
 	}
 
-	setupLog.Info("starting manager")
-	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
-		setupLog.Error(err, "unable to start manager")
-		cancel()
+	logger.Info(fmt.Sprintf("starting %v subscription controller and manager", commander))
```
Comment (marcobebway) [inline components/eventing-controller/cmd/eventing-controller/main.go:71]: The `%v` will use the default format which is the internal struct view.

Comment (marcobebway) [inline components/eventing-controller/cmd/eventing-controller/main.go:71]: The `%v` will show the default format (which is the internal view for the impl struct). Can we improve this please?

File: components/eventing-controller/cmd/eventing-controller/main.go (hunk 1 above)
Comment (themue) [reply components/eventing-controller/cmd/eventing-controller/main.go:21]: Yes, seen `env` package too late. Will change it.

File: components/eventing-controller/README.md [hunk 6]
```diff
@@ -2,11 +2,12 @@
 
 ## Overview
 
-This component contains controllers for various CustomResourceDefinitions related to eventing in Kyma. The following controllers come with this container:
-- [`controller`](https://github.com/kyma-project/kyma/blob/main/components/eventing-controller/cmd/eventing-controller/main.go) which lays down the eventing infrastructure in Business Event Bus (BEB).
-- [`nats-controller`](https://github.com/kyma-project/kyma/blob/main/components/eventing-controller/cmd/eventing-controller-nats/main.go) which lays down the eventing infrastructure in [NATS](https://docs.nats.io/nats-concepts/intro).
+This component contains controllers for various CustomResourceDefinitions related to eventing in Kyma. The following controller come with this container:
```
Comment (alexandra-simeonova) [inline components/eventing-controller/README.md:5]:
```suggestion
This component contains controllers for various CustomResourceDefinitions related to Eventing in Kyma. The following controller comes with this container:
```

File: components/eventing-controller/README.md [hunk 7]
```diff
@@ -2,11 +2,12 @@
...
+
+- [`controller`](https://github.com/kyma-project/kyma/blob/main/components/eventing-controller/cmd/eventing-controller/main.go) which lays down the eventing infrastructure in Business Event Bus (BEB) or [NATS](https://docs.nats.io/nats-concepts/intro).
```
Comment (alexandra-simeonova) [inline components/eventing-controller/README.md:7]:
```suggestion
- [`controller`](https://github.com/kyma-project/kyma/blob/main/components/eventing-controller/cmd/eventing-controller/main.go) which lays down the Eventing infrastructure in Business Event Bus (BEB) or [NATS](https://docs.nats.io/nats-concepts/intro).
```

File: components/eventing-controller/README.md [hunk 8]
```diff
@@ -29,49 +30,35 @@ This component contains controllers for various CustomResourceDefinitions relate
     ```sh
     make deploy-eventing-controller-nats-local
```
Comment (alexandra-simeonova) [inline components/eventing-controller/README.md:31]: In the [installation](https://github.com/kyma-project/kyma/blob/8cfb79ec0b054852b9aaf8cbbe5ec696f76c75ba/components/eventing-controller/README.md#installation) section, we still mention both `controller`  and `nats-controller`, so should `nats-controller` also be removed here?

**Review by marcobebway — COMMENTED** [review]
Since we'll have one image, both deployment should have the same image name:

- https://github.com/themue/kyma/blob/single-binary-event-controller/resources/eventing/charts/controller/values.yaml#L16
- https://github.com/themue/kyma/blob/single-binary-event-controller/resources/eventing/charts/nats-controller/values.yaml#L11

File: components/eventing-controller/cmd/eventing-controller/nats/nats.go (hunk 4 above)
Comment (themue) [reply components/eventing-controller/cmd/eventing-controller/nats/nats.go:79]: Is intended as a note for discussion. So far afaik no issue exists.

File: components/eventing-controller/cmd/eventing-controller/main.go (hunk 5 above)
Comment (themue) [reply components/eventing-controller/cmd/eventing-controller/main.go:71]: Ouch, has been intended to show the chosen backend.

File: components/eventing-controller/README.md (hunk 8 above)
Comment (themue) [reply components/eventing-controller/README.md:31]: Yes, will be, also the Makefile. That's part of the WiP I have to discuss with @radufa .

**Comment by themue** [comment]
> Since we'll have one image, both deployment should have the same image name:
> 
> * https://github.com/themue/kyma/blob/single-binary-event-controller/resources/eventing/charts/controller/values.yaml#L16
> * https://github.com/themue/kyma/blob/single-binary-event-controller/resources/eventing/charts/nats-controller/values.yaml#L11

As far as I understood discussion we'll right now will have two charts but with one binary using the set backend variable. Having only one chart will be one of the next steps.

**Comment by marcobebway** [comment]
> As far as I understood discussion we'll right now will have two charts but with one binary using the set backend variable. Having only one chart will be one of the next steps.

Exactly, for now we need two charts but with the same binary.

**Comment by marcobebway** [comment]
> As far as I understood discussion we'll right now will have two charts but with one binary using the set backend variable. Having only one chart will be one of the next steps.

Exactly, and the two charts should use the same image name `eventing-controller`:
- https://github.com/themue/kyma/blob/single-binary-event-controller/resources/eventing/charts/controller/values.yaml#L16
- https://github.com/themue/kyma/blob/single-binary-event-controller/resources/eventing/charts/nats-controller/values.yaml#L11 (this still has `name: "eventing-controller-nats"`, which should change to `eventing-controller`)

**Comment by themue** [comment]
> > As far as I understood discussion we'll right now will have two charts but with one binary using the set backend variable. Having only one chart will be one of the next steps.
> 
> Exactly, and the two charts should use the same image name `eventing-controller`:
> 
> * https://github.com/themue/kyma/blob/single-binary-event-controller/resources/eventing/charts/controller/values.yaml#L16 (this is correct)
> * https://github.com/themue/kyma/blob/single-binary-event-controller/resources/eventing/charts/nats-controller/values.yaml#L11 (this still has `name: "eventing-controller-nats"`, which should change to `eventing-controller`)

Done.

**Review by radufa — APPROVED** [review]

**Review by alexandra-simeonova — APPROVED** [review]

**Pull request merged**